lint/lint.go

```diff
@@ -97,8 +98,17 @@ type Checker struct {
 
 // Check runs rule checker over file f.
 func (c *Checker) Check(f *ast.File) []Warning {
-	c.ctx.warnings = c.ctx.warnings[:0]
-	c.check(f)
+	if c.ctx.QuickFix {
+		c.ctx.FileMutex.Lock()
+		c.check(f)
+		c.ctx.FileMutex.Unlock()
+	} else {
+		c.ctx.FileMutex.RLock()
+		c.ctx.warnings = c.ctx.warnings[:0]
+		c.check(f)
+		c.ctx.FileMutex.RUnlock()
+
+	}
```

Comment:
If I understand it correctly, this kind of locking will make our code 100% linear (serialized).
First checker will acquire lock and all others will have to wait until it finishes.

Reply:
Correction: not entirely serialized, but it will lead to N parallelism instead of M*N.
Usually, locking should be done before and after and operation (as you do it there), but for the minimal amount of time required. In this program, you want to do lock+unlock around code that modifies the tree:
```go
if c.ctx.QuickFix {
	decl.Type = typ
}
```